### PR TITLE
[Runtime] Update cflags and fix warnings accordingly.

### PR DESCRIPTION
**Pull request by mandel-macaque**
Update flags:

* Add -Werror so that warnings are errors.
* Move to -Wconversion that adds:
  * -Wbitfield-enum-conversion
  * -Wbool-conversion
  * -Wconstant-conversion
  * -Wenum-conversion
  * -Wfloat-conversion
  * -Wimplicit-float-conversion
  * -Wimplicit-int-conversion
  * -Wint-conversion
  * -Wliteral-conversion
  * -Wnon-literal-null-conversion
  * -Wnull-conversion
  * -Wobjc-literal-conversion
  * -Wshorten-64-to-32
  * -Wsign-conversion
  * -Wstring-conversion
Last work related to: https://github.com/xamarin/xamarin-macios/pull/7405

### CODE REVIEW

**Comment by mandel-macaque** [comment]
build

**Comment by spouliot** [comment]
That's bot 38 (first time) and 37 (2nd time) failing 32bits tests

File: tests/test-libraries/libtest.m [hunk 1]
```diff
@@ -730,7 +730,7 @@ -(void) callAssertMainThreadBlockReleaseCallbackQOS
 
 -(void) assertMainThreadBlockReleaseCallback: (innerBlock) completionHandler
 {
-	assert (!"THIS FUNCTION SHOULD BE OVERRIDDEN");
+	assert (false); //!"THIS FUNCTION SHOULD BE OVERRIDDEN");
```
Comment (wachaudh) [inline tests/test-libraries/libtest.m]: minor: remove the `)` in the comment.

Comment (mandel-macaque) [inline tests/test-libraries/libtest.m]: Like this:
```suggestion
	assert (false); // "THIS FUNCTION SHOULD BE OVERRIDDEN"
```

File: tests/test-libraries/libtest.m [hunk 2]
```diff
@@ -602,7 +602,7 @@ + (void)setTestClass:(Class) value {
 
 -(void) classCallback: (void (^)(int32_t magic_number))completionHandler
 {
-	assert (!"THIS FUNCTION SHOULD BE OVERRIDDEN");
+	assert (false); //!"THIS FUNCTION SHOULD BE OVERRIDDEN"
```
Comment (mandel-macaque) [inline tests/test-libraries/libtest.m]:
```suggestion
	assert (false); // THIS FUNCTION SHOULD BE OVERRIDDEN
```

**Comment by mandel-macaque** [comment]
Known issue: https://github.com/xamarin/maccore/issues/581

**Pull request merged**